### PR TITLE
Add state Medicaid ID field to group practices

### LINKED ISSUE #347: Capture Medicaid Number for new Enrollments
# Problem
Requirement psm-2.9 states that PSM application shall accept a medicaid number for new enrollments. There is no field for this value.

Add this field to the ~individual and~ organizational info page~s~ in the enrollment workflow and update the serenity tests to confirm.

### PR DESCRIPTION
**Pull request by jasonaowen**
Some states issue providers a state-specific ID to track their enrollment in the state's Medicaid program. Sometimes providers have multiple NPIs, but in such states will only have a single state Medicaid ID.

Ask providers who do not have a private practice to (optionally) provide this state Medicaid ID, so that they may be linked to the group provider. As we don't know what shape those IDs take, right now it's a free-form text field of up to 100 characters.

Note that this field is only shown for the primary practice. Additional practices have only a short form, which is missing some fields. We can add this field to that form separately, if needed - @cecilia-donnelly, what do you think?

---

screenshot 2017-09-11 16 54 31

---

Deployment: this alters the database. You can either dump and re-seed, or you may open a database connection (ie, with `psql`) and run `ALTER TABLE organizations ADD COLUMN state_medicaid_id TEXT;`.

---

To test, I created new enrollments as individual providers and verified that with a private practice, the field is absent, and with a primary practice, the field is present. The data is persisted between screens (forward/back, and on the summary page), and upon submission is in the database (`SELECT entity_id, state_medicaid_id FROM organizations;`). The admin can edit the enrollment, change the ID, and that is persisted; switching from a primary to a private practice drops the field, and vice versa.

Resolves #347 Capture Medicaid Number for new Enrollments

### CODE REVIEW

**Comment by cecilia-donnelly** [comment]
Thanks a lot for the `psql` instruction.  I was able to submit an enrollment with this ID for a "Community Health Care Worker," yay!  I'm finding the code changes a bit confusing in some places (not your fault) -- still reading.

File: psm-app/db/seed.sql [hunk 1]
```diff
@@ -988,6 +988,7 @@ CREATE TABLE organizations(
   ten99_address_id BIGINT
     REFERENCES addresses(address_id),
   state_tax_id TEXT,
+  state_medicaid_id TEXT,
```
Comment (cecilia-donnelly) [inline psm-app/db/seed.sql]: Ha, I thought you missed this for a sec but you're so on it!

**Comment by jasonaowen** [comment]
> do you mind explaining why we have two files named `primary_practice.jsp`? It looks like one is for entering information and the other is for displaying it -- is that right?

Yes, exactly. The JSP files in `.../default/` are for entering information, and the files in `.../common/` are for displaying it. There's definitely a case to be made for renaming these; I haven't done that yet because it works and has at least a little bit of internal logic which I can understand.

**Comment by jasonaowen** [comment]
> is there a good resource you know of for understanding the `FormBinder`s?

I don't know of any, I'm afraid. The basic idea is to have a way to map from `POST`ed HTML form data to a Java model, and from a Java model to the data the JSPs need. It has a namespacing system (prefixes like `_05_`) to allow multiple forms to be on the same page without interfering with each other.

Spring has built-in support to solve this problem, but I don't know how well it works with multiple, combined forms (or rather fragments of forms), as happens in our app.

As an additional complication, not only do we have both the HTML/form representation and the database representation, we also have the XML representation. I still don't know for sure why it was built that way - my best guess is that it has to do with the jBPM process management, as submitting an enrollment seems to both persist the data to the database as expected, as well as sendingan XML payload to jBPM, which persists it as a blob in its own database tables.

**Comment by jasonaowen** [comment]
Rebased onto master to fix merge conflict with #390. (This is an important reminder not to mix whitespace changes with content changes in my commits; this only ever leads to trouble!)

**Comment by jasonaowen** [comment]
@cecilia-donnelly says:

> If we decide we need the medicaid number anywhere else we can open another issue for it

So, resolving #347 with this PR.

**Pull request merged**